make-release.sh

```diff
@@ -0,0 +1,159 @@
+  echo "    --no-commit, -n"
+  echo "        Do not commit changes."
+  echo "    --trigger-release, -t"
+  echo "        Push changes to remote repo, triggering release."
+  echo "Example: $0 --repo git@github.com:eclipse/che-subproject --version 'v3.2.1' --trigger-release"; echo
+}
+
+while [[ "$#" -gt 0 ]]; do
+  case $1 in
+    '-r'|'--repo') REPO="$2"; shift 1;;
+    '-v'|'--version') VERSION="$2"; shift 1;;
+    '-t'|'--trigger-release')
+      if [[ ! ${NOCOMMIT} -eq 0 ]]; then
+        echo "Only one of --trigger-release and --no-commit may be specified"
+        exit 1
+      fi
+      TRIGGER_RELEASE=1; shift 0;;
+    '-n'|'--no-commit')
+      if [[ ! ${NOCOMMIT} -eq 0 ]]; then
+        echo "Only one of --trigger-release and --no-commit may be specified"
+        exit 1
+      fi
+      NOCOMMIT=1; shift 0;;
+    '-h'|'--help'|*) usage; exit 0;;
+  esac
+  shift 1
+done
+
+if [[ ! ${VERSION} ]] || [[ ! ${REPO} ]]; then
+  usage
+  exit 1
+fi
+
+if [[ ${VERSION} != "v"* ]]; then
+  echo "Version must be prefixed with 'v'"
+  exit 1
+fi
+
+CURRENT_MAJOR_VERSION=$(cut -f 1 -d '.' VERSION)
+NEW_MAJOR_VERSION=$(echo "$VERSION" | cut -f 1 -d '.')
+if [[ "${NEW_MAJOR_VERSION}" != "${CURRENT_MAJOR_VERSION}" ]]; then
+  MAJOR_VERSION_BUMP=1
+fi
+
+# derive branch from version
+BRANCH=${VERSION%.*}.x
+
+# if doing a .0 release, use master; if doing a .z release, use $BRANCH
+if [[ ${VERSION} == *".0" ]]; then
+  BASEBRANCH="master"
+else
+  BASEBRANCH="${BRANCH}"
+fi
+
+# work in tmp dir
+TMP=$(mktemp -d); pushd "$TMP" > /dev/null || exit 1
+
+# get sources from ${BASEBRANCH} branch
+echo "Check out ${REPO} to ${TMP}/${REPO##*/}"
+git clone "${REPO}" -q
+cd "${REPO##*/}" || exit 1
+git fetch origin "${BASEBRANCH}":"${BASEBRANCH}"
+git checkout "${BASEBRANCH}"
+
+# create new branch off ${BASEBRANCH} (or check out latest commits if branch already exists), then push to origin
+if [[ "${BASEBRANCH}" != "${BRANCH}" ]]; then
+  git branch "${BRANCH}" || git checkout "${BRANCH}" && git pull origin "${BRANCH}"
+  git push origin "${BRANCH}"
+  git fetch origin "${BRANCH}:${BRANCH}"
+  git checkout "${BRANCH}"
+fi
+
+# change VERSION file
+echo "${VERSION}" > VERSION
+
+# commit change into branch
+if [[ ${NOCOMMIT} -eq 0 ]]; then
+  COMMIT_MSG="[release] Bump to ${VERSION} in ${BRANCH}"
+  git commit -s -m "${COMMIT_MSG}" VERSION
+  git pull origin "${BRANCH}"
+  git push origin "${BRANCH}"
+fi
+
+if [[ ${TRIGGER_RELEASE} -eq 1 ]]; then
+  # push new branch to release branch to trigger CI build
+  git fetch origin "${BRANCH}:${BRANCH}"
+  git checkout "${BRANCH}"
+  git branch release -f
+  git push origin release -f
+
+  # tag the release
+  git checkout "${BRANCH}"
+  git tag "${VERSION}"
+  git push origin "${VERSION}"
+fi
+
+# now update ${BASEBRANCH} to the new snapshot version
+git fetch origin "${BASEBRANCH}":"${BASEBRANCH}"
+git checkout "${BASEBRANCH}"
+
+# change VERSION file + commit change into ${BASEBRANCH} branch
```

Comment:
I see you've spiffed up my script to add support for a hypothetical 4.0.0 release. Planning more breaking API changes any time soon? :)

Reply:
Always and forever ;)

The broker changes major version relatively often (at least compared to Che server), so I had a choice between "check if major is incremented and fail if so" or "check if major is implemented and handle it". I did option 1 and realized I may as well support it.

I predict container merging will warrant a v4 since it'll change the structure of the output.